File: LeetCode/JavaScript/#404 Sum of Left Leave.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// recursive
var sumOfLeftLeaves = function(root) {
	var sum = 0;
    if(root===null){
    	return 0;
    }
    if(root.left!==null){
    	if(root.left.left===null&&root.left.right===null){
	   		sum +=root.left.val;
	   	}else{
	   		sum+=sumOfLeftLeaves(root.left);
	   	}
    }
    sum+=sumOfLeftLeaves(root.right);
   	return sum;
};

// iterative
var sumOfLeftLeaves = function(root){
	if(root===null) return 0;
	var s = [root];
	var sum = 0;
	var temp;
	while(s.length!==0){
		temp = s.pop();
		if(temp.left!==null){
			if(temp.left.left===null&&temp.left.right===null){
				sum +=temp.left.val;
			}else{
				s.push(temp.left);
			}
		}
		if(temp.right!==null){
			if(temp.right.left===null&&temp.right.right===null){
				s.push(temp.right);
			}
		}
	}
	return sum;
};

File: LeetCode/Python/#404 Sum of Left Leave.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        if root.left and not root.left.left and not root.left.right:
            return root.left.val+self.sumOfLeftLeaves(root.right)
        return self.sumOfLeftLeaves(root.left)+self.sumOfLeftLeaves(root.right)

#
class Solution(object):
    def sumOfLeftLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def helper(root,left=True):
            if not root:return 0

            # root.left is root.right unless both of them are None

            if left and root.left is root.right:
                return root.val
            return helper(root.left,True)+helper(root.right)

        return helper(root,False)